meson: explicitly configurable options

orlitzky opened the pull request:
As discussed in various places over the past two years, it is useful to be able to override the automagic detection of libraries to explicitly enable or disable a feature. This commit adds `meson.options` at the top level to do that, and to make the documentation optional as well.

Example:

```
$ meson setup -Dsirocco=disabled
```

will disable the cython module `sage.libs.sirocco`, even if I have sirocco installed. This is nice for a few reasons:

* If you are building from source (or using a source-based distro), this is the only way to achieve accurate dependency information.
* Now that we are adding meson subprojects, there is again a risk of vendoring libraries that you already have installed if (say) its pkg-config check is flaky. Specifying `-Dfoo=enabled` makes that an error rather than quietly doing the wrong thing.
* It makes it possible to easily test the `# needs whatever` tags on non-Windows systems (i.e. to test that the things we say are optional are really optional).
* On a system where the corresponding libraries happen to be installed, you can speed up the build by disabling the features you aren't currently working on or using.

And of course, the defaults are all "auto," and should not change anything for anyone else.

I was finally motivated to do this because the doc build fails on my machine, and otherwise I have to edit out the line in `src/meson.build` and be careful not to commit it to all of my PRs. This implementation is the first thing I came up with after a few minutes of reading the docs, but it works surprisingly well.

orlitzky committed:
meson: make doc build optional

Make it possible to disable the documentation build by passing,

  $ meson setup -Dbuild-docs=false

Otherwise, the documentation will be built by default, as usual.

orlitzky committed:
meson: make several optional features explicitly configurable

Several libraries are detected at setup time, and the corresponding
feature is enabled or disabled based on the library's presence. This
commit retains this behavior by default while making it possible to
explicitly enable or disable the feature via,

  $ meson setup -Dfoo=disabled

or

  $ meson setup -Dfoo=enabled

The former disables the feature even if the library is installed, and the
latter makes it an error if the library is not installed.

The list of features (valid values for foo) is,

  * bliss
  * brial
  * coxeter3
  * mcqd
  * meataxe
  * rankwidth
  * sirocco
  * tdlib

tobiasdiez commented on src/meson.build:
```diff
@@ -238,4 +246,7 @@ src = meson.current_source_dir()
 
 # Submodules
 subdir('sage')
-subdir('doc')
```
I think it's a tad cleaner if this check is in `doc/meson.build` similar to https://github.com/vbraun/sage/commit/9573c67a62fe4414d1bed1778a8e646d051c892c

orlitzky replied:
Done, I cherry-picked that commit to avoid a merge conflict.

tobiasdiez commented on meson.options:
```diff
@@ -0,0 +1,26 @@
+# Build the HTML / PDF documentation
```
Please add the comments under the `description` field, then they are shown also by the meson cli.

orlitzky replied:
I missed that sentence in the docs. Fixed!

tobiasdiez commented on meson.options:
```diff
@@ -0,0 +1,26 @@
+# Build the HTML / PDF documentation
+option('build-docs', type: 'boolean')
+
+# Build sage.graphs.bliss
```
Not sure if it's more helpful (but perhaps more future-proof) to say "everything that relies on `bliss`" instead of the specific sage module that currently uses it.

orlitzky replied:
Agreed, I tried to make these more sensible now that they're user-facing.

tobiasdiez reviewed (changes requested):
I can certainly see that these options can be helpful in certain situations, and the maintenance overhead is quite small. So I'm fine with going forward with this PR.

If you feel like it, you can also add a bit of documentation on how to toggle those features. Should be enough to point to `meson introspect --buildoptions` for the full list of options (https://mesonbuild.com/Commands.html#introspect).

The CI is failing due to some small formatting issues.

Two small remarks:

> If you are building from source (or using a source-based distro), this is the only way to achieve accurate dependency information.

Not sure if I understand this comment correctly, but the dependency info can be found in `builddir\meson-info\intro-dependencies.json` and in `meson-logs`.

> Now that we are adding meson subprojects, there is again a risk of vendoring libraries that you already have installed if (say) its pkg-config check is flaky. Specifying -Dfoo=enabled makes that an error rather than quietly doing the wrong thing.

In this case, it's better to use `--wrap-mode` (https://mesonbuild.com/Subprojects.html#commandline-options) to skip all submodules.

orlitzky committed:
src/doc/en/installation/meson.rst: document explicit configuration

Add examples of explicit configuration using the "-Doption=value"
syntax, and mention the auto_features option for package maintainers.

antonio-rojas reviewed (approved):
Works fine here, tested no configuration (everything is built) and disabling a couple of modules while still keeping the dependencies installed.

dimpase reviewed (commented):
how about mentioning `--wrap-mode=...` in the docs?

orlitzky committed:
src/doc/en/installation/meson.rst: mention the --wrap-mode flag

Mention how to disabled wrapped subprojects in the tips for package
maintainers section.

orlitzky committed:
src/sage/libs/**/meson.build: wrap a few long lines

tobiasdiez reviewed (approved):
Okay, then let's get this in!

orlitzky committed:
pyproject.toml: move brial and eclib to optional-host-requires

It is now possible to build sagelib without these libraries; the
corresponding extensions are automatically disabled by meson.

orlitzky committed:
src/meson.build: require eclib >= 20250122

I believe this is required anyway after commit eaef4ae0, and
regardless is nice to have because it guarantees that we'll have a
pkg-config file and can skip the no-pkgconfig fallback.

orlitzky committed:
meson: add explicit option for eclib support

Add an entry to meson.options (and tweak src/meson.build) so that
the user can explicitly state whether or not he would like eclib
support to be enabled, e.g.

  $ meson setup -Declib=disabled

The default of "auto" is backwards-compatible.

orlitzky committed:
meson.options: improve the description of the "meataxe" feature

orlitzky committed:
src/meson.build: run tools/update-meson.py to add a comma
